### PR TITLE
#2664 GeneratorPVFixed added, LoadFlow model of Nordic 32 uses it

### PR DESCRIPTION
**Pull request by ErwanGuichard49**
closes [2664](https://github.com/dynawo/dynawo/issues/2664)

### CODE REVIEW

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/Simplified/GeneratorPVFixed.extvar [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, you can obtain one at http://mozilla.org/MPL/2.0/.
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source suite
+    of simulation tools for power systems.
+-->
+<external_variables xmlns="http://www.rte-france.com/dynawo">
+  <variable type="continuous" id="terminal.V.re"/>
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/Simplified/GeneratorPVFixed.extvar:15]: We impose V in this model, so for me it should be i here

File: dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/Simplified/GeneratorPVFixed.mo [hunk 2]
```diff
@@ -0,0 +1,36 @@
+within Dynawo.Electrical.Machines.Simplified;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Electrical/Machines/Simplified/GeneratorPVFixed.mo:1]: Can you create a preassembled model for GeneratorPVFixed ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo [hunk 3]
```diff
@@ -135,296 +135,335 @@ model NetworkWithAlphaBetaLoads "Nordic test grid with buses, lines, shunts and
...
-  final parameter Types.ComplexPerUnit u0Pu_Load_72 = ComplexMath.fromPolar(U0Pu_Load_72, UPhase0_Load_72);
-  final parameter Types.ComplexPerUnit i0Pu_Load_72 = ComplexMath.conj(s0Pu_Load_72 / u0Pu_Load_72);
-  final parameter SIunits.Impedance XBase_130 = 130 ^ 2 / Electrical.SystemBase.SnRef;
-  final parameter SIunits.Impedance XBase_220 = 220 ^ 2 / Electrical.SystemBase.SnRef;
-  final parameter SIunits.Impedance XBase_400 = 400 ^ 2 / Electrical.SystemBase.SnRef;
+  // load_01 init values:
+  // s0Pu, i0Pu in SnRef, receptor convention
+  final parameter Types.ActivePowerPu P0Pu_load_01 = 6;
+  final parameter Types.ReactivePowerPu Q0Pu_load_01 = 1.482;
+  final parameter Types.VoltageModulePu U0Pu_load_01 = 0.9988;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo:142]: You need to make sure that this values (U0Pu) are updated for case B as they determine the behaviour of the alpha-beta loads ! So maybe a final parameter is not ideal here ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo [hunk 4]
```diff
@@ -135,296 +135,335 @@ model NetworkWithAlphaBetaLoads "Nordic test grid with buses, lines, shunts and
...
+  final parameter Types.Angle UPhase0_load_01 = SIunits.Conversions.from_deg(-84.71);
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo:143]: Same here

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo [hunk 5]
```diff
@@ -135,296 +135,335 @@ model NetworkWithAlphaBetaLoads "Nordic test grid with buses, lines, shunts and
...
   // shunt_1022 init values:
   // negative values for capacitors, positive values for inductors (reversed in PESTR)
-  final parameter Types.PerUnit BPu_shunt_1022 = -50.0 / Electrical.SystemBase.SnRef;
-  final parameter Types.ComplexPerUnit u0Pu_shunt_1022 = ComplexMath.fromPolar(1.0512, SIunits.Conversions.from_deg(-19.05));
-  final parameter Types.PerUnit U0Pu_shunt_1022 = ComplexMath.'abs'(u0Pu_shunt_1022);
-  final parameter Types.PerUnit Q0Pu_shunt_1022 = BPu_shunt_1022 * U0Pu_shunt_1022 ^ 2;
-  final parameter Types.ComplexPerUnit s0Pu_shunt_1022 = Complex(0, Q0Pu_shunt_1022);
-  final parameter Types.ComplexPerUnit i0Pu_shunt_1022 = ComplexMath.conj(s0Pu_shunt_1022 / u0Pu_shunt_1022);
+  final parameter Types.PerUnit BPu_shunt_1022 = -0.5;
+  final parameter Types.VoltageModulePu U0Pu_shunt_1022 = 1.0512;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo:361]: Should be updated for case B

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo [hunk 6]
```diff
@@ -135,296 +135,335 @@ model NetworkWithAlphaBetaLoads "Nordic test grid with buses, lines, shunts and
...
+  final parameter Types.Angle UPhase0_shunt_1022 = SIunits.Conversions.from_deg(-19.05);
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo:362]: same

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullDynamicModel.mo [hunk 7]
```diff
@@ -169,84 +169,103 @@ model FullDynamicModel "Nordic test grid with buses, lines, shunts, voltage-depe
 protected
   // Generator g01 init values:
   // P0Pu, Q0Pu in SnRef, receptor convention
-  final parameter Types.PerUnit P0Pu_g01 = -600.00 / Electrical.SystemBase.SnRef;
-  final parameter Types.PerUnit Q0Pu_g01 = -58.3 / Electrical.SystemBase.SnRef;
+  final parameter Types.ActivePowerPu P0Pu_g01 = -6;
+  final parameter Types.ReactivePowerPu Q0Pu_g01 = -0.583;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullDynamicModel.mo:173]: The Q0Pu here need to be taken from the load flow and updated for case B, so maybe not a final parameter ?

**Review by cossartque — CHANGES_REQUESTED** [review]
Couldn't we remove all numerical init values from NetworkWithAlphaBetaLoads, NetworkWithPQLoads and FullDynamicModel, to give them all directly in LoadFlow and TestCase ? This would avoid some mistakes + the only files to change for case B would be LoadFlow and TestCase

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo [hunk 8]
```diff
@@ -14,417 +14,311 @@ within Dynawo.Examples.Nordic.Grid.BaseClasses;
...
-  final parameter Types.Angle UPhase0_Load_72 = SIunits.Conversions.from_deg(-6.83);
-  final parameter Types.VoltageModule U0Pu_Load_72 = 0.9974;
-  final parameter Types.ComplexPerUnit u0Pu_Load_72 = ComplexMath.fromPolar(U0Pu_Load_72, UPhase0_Load_72);
-  final parameter Types.ComplexPerUnit i0Pu_Load_72 = ComplexMath.conj(s0Pu_Load_72 / u0Pu_Load_72);
-  final parameter SIunits.Impedance XBase_130 = 130 ^ 2 / Electrical.SystemBase.SnRef;
-  final parameter SIunits.Impedance XBase_220 = 220 ^ 2 / Electrical.SystemBase.SnRef;
-  final parameter SIunits.Impedance XBase_400 = 400 ^ 2 / Electrical.SystemBase.SnRef;
+  // load_01 init values:
+  // s0Pu, i0Pu in SnRef, receptor convention
+  parameter Types.ActivePowerPu P0Pu_load_01 = 6;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo:138]: Why didn't you remove the numerical values here too (like you did for alpha beta)?

Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo:138]: And then give the values in LoadFlow

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite
+* of simulation tools for power systems.
+*/
+
+model FullStaticModel "Nordic test grid with buses, lines, shunts, loads, transformers and generators, used for load flow calculation"
+  import Modelica.ComplexMath;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo:17]: Not used ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo [hunk 10]
```diff
@@ -14,417 +14,311 @@ within Dynawo.Examples.Nordic.Grid.BaseClasses;
...
+  Electrical.Shunts.ShuntB shunt_4046(BPu = BPu_shunt_4046, u0Pu = Complex(1, 0), s0Pu = s0Pu_shunt_4046, i0Pu = s0Pu_shunt_4046) annotation(
     Placement(visible = true, transformation(origin = {76.5, -22.5}, extent = {{-2.5, 2.5}, {2.5, -2.5}}, rotation = 0)));
-  Electrical.Shunts.ShuntB shunt_4051(BPu = BPu_shunt_4051, u0Pu = u0Pu_shunt_4051, s0Pu = s0Pu_shunt_4051, i0Pu = i0Pu_shunt_4051) annotation(
+  Electrical.Shunts.ShuntB shunt_4051(BPu = BPu_shunt_4051, u0Pu = Complex(1, 0), s0Pu = s0Pu_shunt_4051, i0Pu = s0Pu_shunt_4051) annotation(
     Placement(visible = true, transformation(origin = {21.5, -134.5}, extent = {{-2.5, -2.5}, {2.5, 2.5}}, rotation = 0)));
-  Electrical.Shunts.ShuntB shunt_4071(BPu = BPu_shunt_4071, u0Pu = u0Pu_shunt_4071, s0Pu = s0Pu_shunt_4071, i0Pu = i0Pu_shunt_4071) annotation(
+  Electrical.Shunts.ShuntB shunt_4071(BPu = BPu_shunt_4071, u0Pu = Complex(1, 0), s0Pu = s0Pu_shunt_4071, i0Pu = s0Pu_shunt_4071) annotation(
     Placement(visible = true, transformation(origin = {-85.5, 136.5}, extent = {{2.5, 2.5}, {-2.5, -2.5}}, rotation = 0)));
 
 protected
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo:135]: We can remove protected

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo (hunk 8 above)
Comment (cossartque) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo:138]: (same for shunts)

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo [hunk 11]
```diff
@@ -15,101 +15,100 @@ within Dynawo.Examples.Nordic.Grid.BaseClasses;
...
 model NetworkWithAlphaBetaLoads "Nordic test grid with buses, lines, shunts and voltage-dependent loads"
   import Modelica.ComplexMath;
-  import Modelica.SIunits;
   import Dynawo;
   import Dynawo.Electrical;
   import Dynawo.Types;
 
   extends Dynawo.Examples.Nordic.Grid.BaseClasses.Network;
 
-  Electrical.Loads.LoadAlphaBeta load_01(alpha = 1, beta = 2, s0Pu = s0Pu_Load_01, u0Pu = u0Pu_Load_01, i0Pu = i0Pu_Load_01) annotation(
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo:25]: You can have a generic parameter alpha and a generic parameter beta to be used for all the loads. alpha = 1 and beta = 2 will be given in the test case

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+  import Dynawo;
+  import Dynawo.Electrical;
+  import Dynawo.Electrical.SystemBase;
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo:20]: Not used ?

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo [hunk 13]
```diff
@@ -0,0 +1,692 @@
...
+    Placement(visible = true, transformation(origin = {-78, -88}, extent = {{-3, -3}, {3, 3}}, rotation = 0)));
+  Electrical.Machines.Simplified.GeneratorPVFixed g18(PGen0Pu = P0Pu_g18, U0Pu = U0Pu_g18) annotation(
+    Placement(visible = true, transformation(origin = {-95, -151}, extent = {{-3, -3}, {3, 3}}, rotation = 0)));
+  Electrical.Machines.Simplified.GeneratorPVFixed g19(PGen0Pu = P0Pu_g19, U0Pu = U0Pu_g19) annotation(
+    Placement(visible = true, transformation(origin = {-75, 151}, extent = {{-3, -3}, {3, 3}}, rotation = 0)));
+
+protected
+  // g01 init values:
+  // P0Pu, Q0Pu in SnRef, generator convention
+  // i0Pu in receptor convention
```
Comment (cossartque) [inline dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo:170]: not needed (for all generators)

**Review by cossartque — CHANGES_REQUESTED** [review]
A few comments remaining:
- We want to avoid all numerical values in the "Grid" repo, except for X, G, B, R, rTfo.
- Please check that all the imports and used, otherwise we can remove them
- We can remove all the "protected"

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo (hunk 9 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo:17]: Removed

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo (hunk 10 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo:135]: Removed

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo (hunk 12 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo:20]: Deleted

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo (hunk 13 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/FullStaticModel.mo:170]: Lines deleted

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo (hunk 11 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithAlphaBetaLoads.mo:25]: Done

File: dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo (hunk 8 above)
Comment (ErwanGuichard49) [reply dynawo/sources/Models/Modelica/Dynawo/Examples/Nordic/Grid/BaseClasses/NetworkWithPQLoads.mo]: Done

**Review by cossartque — APPROVED** [review]

**Pull request merged**